HHS/TANF-app
# Backend/309/dynamic roles permissions

### This pull request changes...

- Adds CRUD endpoints for roles and permissions
- Adds starting roles to the database

## FUNCTIONALITY
The backend is now fully functional for creating and retrieving roles and permissions

## TO TEST

*Note*: An API client is needed to test, preferably one that works as a chrome extension so it can ready your HTTPOnly cookies.

1. Go to the [Cloud Dev Site](https://tdp-frontend.app.cloud.gov/)
2. Click the "Sign In" button to login through login.gov
3. Once successfully logged in, open your API Client
4. Paste the following url into the request field: https://tdp-backend.app.cloud.gov/v1/roles/ 
5. Ensure the `GET` method is selected
6. Click `SEND` (or similar in different API Clients)

If the return body looks like this, the test passes:
```
[
    {
        "id": 1,
        "name": "OFA Admin",
        "permissions": []
    },
    {
        "id": 2,
        "name": "Data Prepper",
        "permissions": []
    },
    {
        "id": 3,
        "name": "OFA Analyst",
        "permissions": []
    }
]
```

### This pull request is ready to merge when...

  - OpenAPI  
  - Readme
  - Entity Relationship Diagram (ERD)

**Including the QASP Checklist for our initial review**

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [ ] Preliminary (MVP) roles are populated in the database

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: 100% (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [ ] Backend coverage: 100% (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*

    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/a11y/how-18f-will-test-a11y.md)

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful build with a single command*


### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*


### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*
